study_guide/questions.md

```diff
@@ -0,0 +1,179 @@
+- What's the "short string optimization"?
+
+== Composite data types. Class, Structs and unions.
+
+- What's a class? What's a structure? What's a union?
+- What are the similarities and differences between `class` and `struct`?
+- Can a class inherit? Can a struct inherit?
+- What kinds of members can a class have?
+- What are access specifiers?
+- What's the difference between private and protected?
+- What's `const` method of a class?
+- What's `this`? 
+- What's a static member of a class?
+
+== Enum class vs enum.
+
+- What's an `enum`? In what name scope are the constants created?
+- What's a `class enum`? In what name scope are the constants created?
+- How does `enum` compare against `enum class`?
+
+== References and pointers. Primitive arrays.
+
+- What a raw pointer?
+- How do you create a raw pointer? How do you release a raw pointer?
+- What'a reference? How do you create a reference?
+- What are the similarities and differences between a reference and a pointer?
+- What's a dangling reference? provide a code example that creates a dangling reference.
+- what's "void"? What can you do with it?
+- What are the advantages of passing parameters to functions by `const T &`?
+- What's the rule of thumb to determine when to pass parameters to functions by `const T &`?
+
+== Const and constexpr. volatile.
+
+- What does it mean to declare a variable as `const`?
+- Is `const` a build-time or run-time feature?
+- Where is a `const` variables stored?
+- Can I create a `const` variable that holds an information piece the user provided during execution? (i.e: through the terminal).
+- What does it mean to declare a variable as `constexpr`?
+- Is `constexpr` a build-time or run-time feature?
+- Where are `constexpr` variables stored?
+- Can I create a `constexpr` variable that holds an information piece the user provided during execution? (i.e: through the terminal).
+- What does it mean for a function to be `constexpr`?
+- What limitations does a `constexpr` function has?
+- What does `volatile` do? Explain the difference in the generated assembler code in https://godbolt.org/z/sac448 .
+
+== “Auto” statement. Uniform initialization. 
+
+- What does the `auto` statement does?
+- What are the advantages of using `auto`?
+- How does `auto` interact with the uniform initializer `{}`?
+
+== Namespaces. Anonymous namespaces.
+
+- What's a namespace?
+- What do you use a namespace for?
+- What namespaces do you know?
+- What's an anonymous namespace? What is it for?
+- What does the `using namespace ...` clause do? What problem can it bring?
+
+== Life scope of variables: automatic, global, dynamic.
+
+- What's the life scope of a variable?
+- What's the scope of a global variable? where is it stored?
+- What's an automatic variable? Where is it stored?
+- What's the scope of a variable in dynamic memory? where is it stored?
+
+== Construction and destruction:
+
+- What is a class constructor?
+- How many constructors are there in a class?
```

Comment:
Technically, a class could allocate an unlimited number of constructors. 
I guess you are referring to the "typical" constructors, right?

Reply:
Actually it was meant as a "trick" question, and yes, the answer was "potentially infinite", but that's probably not really helpful for a padawan. I'll rewrite it, thanks.